File: TypeScript/reverse-complement/revcomp.js
/* The Computer Language Benchmarks Game
   http://benchmarksgame.alioth.debian.org/

   node.js version by Joe Farro
   TypeScript adaptation by Josh Goldfoot
*/
"use strict";
/// <reference path="../node_modules/@types/node/index.d.ts" />
class LinkedArray {
    constructor(prev) {
        this.prev = prev;
        this.next = null;
        this.pos = 0;
        this.data = [];
    }
}
const stdout = process.stdout;
const stdin = process.stdin;
const READ_SIZE = 16000;
const writeBuffer = new Buffer(READ_SIZE + READ_SIZE / 61 | 0);
var smap = [, , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , ,
    ,
    , , , , 84, 86, 71, 72, , , 67, 68, , , 77, , 75, 78, , , , 89, 83, 65, 65, 66, 87, , 82, , , , , , , ,
    84, 86, 71, 72, , , 67, 68, , , 77, , 75, 78, , , , 89, 83, 65, 65, 66, 87, , 82];
const LA_LEN = 30;
const headLA = new LinkedArray(null);
let la = headLA;
let lnIdx = 0;
let lines = la.data;
let needHeader = true;
let headerPartial = '';
let isFirst = true;
var metaI;
var numLines;
function read() {
    let chunk = stdin.read(READ_SIZE);
    let isFinal = false;
    if (!chunk) {
        if (isFirst) {
            isFirst = false;
            return;
        }
        la.pos = lnIdx;
        reverse(la);
        return;
    }
    else if (chunk.length < READ_SIZE) {
        isFinal = true;
    }
    while (chunk) {
        while (true) {
            // if have read a partial header line, read the rest of it
            if (needHeader) {
                const headerEnds = chunk.indexOf('\n');
                console.log(headerPartial.toString() + chunk.slice(0, headerEnds).toString('ascii'));
                headerPartial = '';
                chunk = chunk.slice(headerEnds);
                needHeader = false;
            }
            const caretIdx = chunk.indexOf('>');
            if (caretIdx > -1) {
                // there is a caret in this chunk -- process the first part of
                // the chunk and then continue the `while (true)` loop to process
                // the next part of the chunk
                lines[lnIdx] = chunk.slice(0, caretIdx);
                // set chunk to the next section
                chunk = chunk.slice(caretIdx);
                la.pos = lnIdx + 1;
                reverse(la);
                // reset the data holders
                la = headLA;
                lines = la.data;
                la.pos = 0;
                lnIdx = 0;
                // check to see if the entire header line is here
                const headerEnds = chunk.indexOf('\n');
                if (headerEnds < 0) {
                    needHeader = true;
                    headerPartial = chunk;
                    break;
                }
                else {
                    // log the header
                    console.log(chunk.slice(0, headerEnds).toString('ascii'));
                    // continue processing the rest of the chunk
                    chunk = chunk.slice(headerEnds);
                }
            }
            else {
                if (lnIdx === LA_LEN) {
                    la.pos = LA_LEN;
                    la = la.next || (la.next = new LinkedArray(la));
                    lines = la.data;
                    lines[0] = chunk;
                    lnIdx = la.pos = 1;
                }
                else {
                    lines[lnIdx] = chunk;
                    lnIdx++;
                }
                break;
            }
        }
        chunk = stdin.read(READ_SIZE);
    }
    if (isFinal) {
        la.pos = lnIdx;
        reverse(la);
    }
}
function reverse(la_) {
    // reset the metaI and numLines in this section
    metaI = 0;
    numLines = 0;
    var la = la_, lines = la.data, lnIdx = la.pos - 1, line = lines[lnIdx];
    while (true) {
        reverseCompPrint(line);
        lnIdx--;
        line = lines[lnIdx];
        if (line) {
            continue;
        }
        la = la.prev;
        if (la === null) {
            break;
        }
        lines = la.data;
        lnIdx = la.pos;
        lnIdx--;
        line = lines[lnIdx];
    }
    if ((metaI - numLines) % 60 !== 0) {
        stdout.write('\n');
    }
}
function reverseCompPrint(line) {
    let _metaI = metaI;
    let _numLines = numLines;
    let count = 0;
    const target = writeBuffer;
    const len = line.length;
    const right = line.length - 1;
    let ileft = 0;
    let iright = 0;
    let c;
    while (iright < len) {
        c = line[right - iright];
        iright++;
        if (c == 10) {
            // skip linebreaks
            if (iright === len) {
                break;
            }
            c = line[right - iright];
            iright++;
        }
        target[ileft] = smap[c];
        ileft++;
        count++;
        if ((count + _metaI - _numLines) % 60 === 0) {
            // need a linebreak
            target[ileft] = 10;
            ileft++;
            count++;
            _numLines++;
        }
    }
    metaI = _metaI + count;
    numLines = _numLines;
    stdout.write(target.slice(0, count).toString('ascii'));
}
stdin.on('readable', read);

File: Rust/reverse-complement/revcomp.rs
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// contributed by the Rust Project Developers
// contributed by Cristi Cobzarenco
// contributed by TeXitoi
// contributed by Matt Brubeck

extern crate rayon;

use std::io::{BufRead, BufReader, Write};
use std::{cmp, io};
use std::fs::File;
use std::mem::replace;

/// This controls the size of reads from the input. Chosen to match the C entry.
const READ_SIZE: usize = 16 * 1024;

/// Lookup table to find the complement of a single FASTA code.
fn build_table() -> [u8; 256] {
    let mut table = [0; 256];
    for (i, x) in table.iter_mut().enumerate() {
        *x = match i as u8 as char {
            'A' | 'a' => 'T',
            'C' | 'c' => 'G',
            'G' | 'g' => 'C',
            'T' | 't' => 'A',
            'U' | 'u' => 'A',
            'M' | 'm' => 'K',
            'R' | 'r' => 'Y',
            'W' | 'w' => 'W',
            'S' | 's' => 'S',
            'Y' | 'y' => 'R',
            'K' | 'k' => 'M',
            'V' | 'v' => 'B',
            'H' | 'h' => 'D',
            'D' | 'd' => 'H',
            'B' | 'b' => 'V',
            'N' | 'n' => 'N',
            i => i,
        } as u8;
    }
    table
}

/// Utilities for splitting chunks off of slices.
trait SplitOff {
    fn split_off_left(&mut self, n: usize) -> Self;
    fn split_off_right(&mut self, n: usize) -> Self;
}
impl<'a, T> SplitOff for &'a mut [T] {
    /// Split the left `n` items from self and return them as a separate slice.
    fn split_off_left(&mut self, n: usize) -> Self {
        let n = cmp::min(self.len(), n);
        let data = replace(self, &mut []);
        let (left, data) = data.split_at_mut(n);
        *self = data;
        left
    }
    /// Split the right `n` items from self and return them as a separate slice.
    fn split_off_right(&mut self, n: usize) -> Self {
        let len = self.len();
        let n = cmp::min(len, n);
        let data = replace(self, &mut []);
        let (data, right) = data.split_at_mut(len - n);
        *self = data;
        right
    }
}

/// Length of a normal line including the terminating \n.
const LINE_LEN: usize = 61;
const SEQUENTIAL_SIZE: usize = 2048;

/// Compute the reverse complement for two contiguous chunks without line breaks.
fn reverse_chunks(left: &mut [u8], right: &mut [u8], table: &[u8; 256]) {
    for (x, y) in left.iter_mut().zip(right.iter_mut().rev()) {
        *y = table[replace(x, table[*y as usize]) as usize];
    }
}

/// Compute the reverse complement on chunks from opposite ends of a sequence.
///
/// `left` must start at the beginning of a line. If there are an odd number of
/// bytes, `right` will initially be 1 byte longer than `left`; otherwise they
/// will have equal lengths.
fn reverse_complement_left_right(mut left: &mut [u8],
                                 mut right: &mut [u8],
                                 trailing_len: usize,
                                 table: &[u8; 256]) {
    let len = left.len();
    if len <= SEQUENTIAL_SIZE {
        // Each iteration swaps one line from the start of the sequence with one
        // from the end.
        while left.len() > 0  || right.len() > 0 {
            // Get the chunk up to the newline in `right`.
            let mut a = left.split_off_left(trailing_len);
            let mut b = right.split_off_right(trailing_len);
            right.split_off_right(1); // Skip the newline in `right`.

            // If we've reached the middle of the sequence here and there is an
            // odd number of bytes remaining, the odd one will be on the right.
            if b.len() > a.len() {
                let mid = b.split_off_left(1);
                mid[0] = table[mid[0] as usize];
            }

            reverse_chunks(a, b, table);

            // Get the chunk up to the newline in `left`.
            let n = LINE_LEN - 1 - trailing_len;
            a = left.split_off_left(n);
            b = right.split_off_right(n);
            left.split_off_left(1); // Skip the newline in `left`.

            // If we've reached the middle of the sequence and there is an odd
            // number of bytes remaining, the odd one will now be on the left.
            if a.len() > b.len() {
                let mid = a.split_off_right(1);
                mid[0] = table[mid[0] as usize]
            }

            reverse_chunks(a, b, table);
        }
    } else {
        let line_count = len / LINE_LEN;
        let mid = line_count / 2 * LINE_LEN; // Split on a whole number of lines.

        let left1 = left.split_off_left(mid);
        let right1 = right.split_off_right(mid);
        rayon::join(|| reverse_complement_left_right(left,  right,  trailing_len, table),
                    || reverse_complement_left_right(left1, right1, trailing_len, table));
    }
}

/// Compute the reverse complement of one sequence.
fn reverse_complement(seq: &mut [u8], table: &[u8; 256]) {
    let len = seq.len() - 1;
    let seq = &mut seq[..len]; // Drop the last newline
    let trailing_len = len % LINE_LEN;
    let (left, right) = seq.split_at_mut(len / 2);
    reverse_complement_left_right(left, right, trailing_len, table);
}

fn run() -> io::Result<()> {
    let stdin = File::open("/dev/stdin")?;
    let size = stdin.metadata()?.len() as usize;
    let mut input = BufReader::with_capacity(READ_SIZE, stdin);
    let mut buf = Vec::with_capacity(size);


    let mut seqs = vec![];
    loop {
        // Read the header line.
        input.read_until(b'\n', &mut buf)?;
        let seq_start = buf.len();
        // Read sequence data.
        input.read_until(b'>', &mut buf)?;
        let i = buf.len() - 1;
        if buf[i] == b'>' {
            // Found the start of a new sequence.
            seqs.push(seq_start..i);
        } else {
            // Reached the end of the input.
            seqs.push(seq_start..buf.len());
            break
        }
    }

    // Compute the reverse complements of each sequence.
    let table = build_table();
    for seq in seqs {
        reverse_complement(&mut buf[seq], &table);
    }

    // Print the result.
    io::stdout().write_all(&buf)
}

fn main() {
    run().unwrap()
}